File: 350_intersection_of_two_arrays_ii/intersection_of_two_arrays_ii.go
package main

import "fmt"

type Solution struct{}

func (s *Solution) intersectionTwoArraysII(nums1 []int, nums2 []int) []int {
	count := make(map[int]int)
	for _, num := range nums1 {
		count[num]++
	}

	var result []int
	for _, num := range nums2 {
		if count[num] > 0 {
			result = append(result, num)
			count[num]--
		}
	}
	return result
}

func main() {
	nums1 := []int{4, 9, 5}
	nums2 := []int{9, 4, 9, 8, 4}
	s := Solution{}
	fmt.Println(s.intersectionTwoArraysII(nums1, nums2))
}

File: 350_intersection_of_two_arrays_ii/intersection_of_two_arrays_ii.py
from collections import Counter
from typing import List

class Solution:
    def intersection_two_arrays_ii(self, nums1: List[int], nums2: List[int]) -> List[int]:
        count1 = Counter(nums1)
        result = []

        for num in nums2:
            if count1[num] > 0:
                result.append(num)
                count1[num] -= 1

        return result


if __name__ == "__main__":
    nums1 = [4, 9, 5]
    nums2 = [9, 4, 9, 8, 4]
    solution = Solution()
    print(solution.intersection_two_arrays_ii(nums1, nums2))
